samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManager.java

```diff
@@ -245,8 +256,19 @@ public ContainerStorageManager(
         containerChangelogSystems, systemFactories, config, this.samzaContainerMetrics.registry());
     this.storeConsumers = createStoreIndexedMap(this.changelogSystemStreams, storeSystemConsumers);
 
+    // TODO HIGH dchen tune based on observed concurrency
+    JobConfig jobConfig = new JobConfig(config);
+    int restoreThreadPoolSize =
+        Math.min(
+            Math.max(containerModel.getTasks().size() * 2, jobConfig.getRestoreThreadPoolSize()),
```

Comment:
Fyi this is the temporary solution to resolve the potential deadlock (2x task count)

Reply:
Follow up for this: https://issues.apache.org/jira/browse/SAMZA-2656